### PR TITLE
Iprox for l0 and l1

### PR DESCRIPTION
**Pull request by MohamedLaghdafHABIBOULLAH**
Add iprox for unconstrained L0 and L1

### CODE REVIEW

File: src/shiftedNormL0.jl [hunk 1]
```diff
@@ -52,3 +52,27 @@ function prox!(
...
+# variable change v = x + s + y:
+# arg min vᵀDv/2 - fᵀv + λ h(v)
+# with fᵢ = gᵢ + dᵢ(xᵢ + sᵢ)
+function iprox!(
+  y::AbstractVector{R},
+  ψ::ShiftedNormL0{R, V0, V1, V2},
+  g::AbstractVector{R},
+  d::AbstractVector{R},
+) where {R <: Real, V0 <: AbstractVector{R}, V1 <: AbstractVector{R}, V2 <: AbstractVector{R}}
+
```
Comment (geoffroyleconte) [inline src/shiftedNormL0.jl:66]: Maybe check that $$d > 0$$.

File: test/partial_prox.jl [hunk 2]
```diff
@@ -54,5 +54,20 @@ for op ∈ (:NormL0, :NormL1, :RootNormLhalf)
         end
       end
     end
+
+    # tests iprox without bounds
+    if op == :NormL0 || op == :NormL1
+      for d ∈ [ones(n), 2 * ones(n), zeros(n)]
```
Comment (geoffroyleconte) [inline test/partial_prox.jl:60]: Hi @MohamedLaghdafHABIBOULLAH , this raises an error with your `iprox!` when `d = zeros(n)`, you should remove it here.
You can test the error in the unit tests when `d = zeros(n)` using this for example:
https://docs.julialang.org/en/v1/stdlib/Test/#Test.@test_throws

**Review by geoffroyleconte — CHANGES_REQUESTED** [review]

File: src/shiftedNormL1.jl [hunk 3]
```diff
@@ -51,3 +51,23 @@ function prox!(
...
+function iprox!(
+  y::AbstractVector{R},
+  ψ::ShiftedNormL1{R, V0, V1, V2},
+  g::AbstractVector{R},
+  d::AbstractVector{R},
+) where {R <: Real, V0 <: AbstractVector{R}, V1 <: AbstractVector{R}, V2 <: AbstractVector{R}}
+  y .= -ψ.xk .- ψ.sj
+
+  for i ∈ eachindex(y)
+    @assert d[i] != 0
```
Comment (geoffroyleconte) [inline src/shiftedNormL1.jl:68]: Does it work with `d[i] < 0`?

Comment (MohamedLaghdafHABIBOULLAH) [reply src/shiftedNormL1.jl:68]: No it does not, it will diverge to $$-\infty$$

**Review by geoffroyleconte — APPROVED** [review]

**Pull request merged**